### ISSUE TITLE
[Outreachy 2023] Improve user experience on early system boot failures

### ISSUE THREAD

**Issue by bluca** [systemd/systemd#26586]
### Component

systemd/journald

### Project description

Improve the user experience when something goes wrong early at system boot. We want to ensure that users are able to debug early boot failures by providing QR-encoded URLs pointing to useful description of errors that the system just experienced.

We also want to ensure that the system meets a minimum baseline level before it is allowed to continue booting, which for this project consists in checking that the battery level is adequate and above a minimum configurable threshold.

Implementing this project will require working on low-level components such as the systemd system manager, the journald logging service, and adding new early boot services that handle virtual terminals, consoles and QR encoding.

### Tasks

- detect battery level in initrd and if low refuse continuing to boot, print message and shut down
- if boot fails for any other reason, display a QR code with a link to the journal message catalog explaining the error that can be scanned on a phone
  - audit error level logs that are relevant for early boot failures, and bump log level up
  - for each such log messages, have an entry in the message catalog, and that the entry has an URL
  - add a service that the journal can activate and that will encode the log message URL into a QR code and display it on the screen for the user
  - the new service should be implemented with socket connection activation: https://github.com/systemd/systemd/issues/26586#issuecomment-1443919239
  - extend the boot assessment functionality to plug in this functionality by for example having a timing-out target

**Comment by poettering** [comment]
implementation idea for "activation-by-journald":

```
* journald: add varlink service that allows subscribing to certain log events,
  for example matching by message ID, or log level returns a list of journal
  cursors as they happen.

* In .socket units, add ConnectStream=, ConnectDatagram=,
  ConnectSequentialPacket= that create a socket, and then *connect to* rather than
  listen on some socket. Then, add a new setting WriteData= that takes some
  base64 data that systemd will write into the socket early on. This can then
  be used to create connections to arbitrary services and issue requests into
  them, as long as the data is static. This can then be combined with the
  aforementioned journald subscription varlink service, to enable
  activation-by-message id and similar.
```

(from the TODO file)

**Comment by 1awesomeJ** [comment]
Good evening @bluca.

I'm extremely elated to have gotten this opportunity to be your protege over the next months.

I know it can be quite demanding on your schedule, but please, I'll like to start my mentorship immediately.


You mostly as well as the other team members here have made very huge impact on my tech journey this year sir.

Please find the time in your schedule-no matter how little it is that I get- to start with me immediately sir.


Thank you so so much. God bless you sir.

**Comment by 1awesomeJ** [comment]
Second.

I really do not understand PID1.

I'm thinking src/core/main.c is what I should be studying sir.

Is that right?

**Comment by 1awesomeJ** [comment]
Third.

What other modules should I start with, in order to get the actual flow of the init process?


Thank you sir.

**Comment by 1awesomeJ** [comment]
going through console_setup() function, I saw a comment that says we don't want to enforce text-mode on the console as initrd may have started a graphical process like plymouth.

up until now, I thought the boot process flow was:
1. Bios/UEFI
2. Grub/Systemd-boot
3. kernel
4. Systemd (PID 1)

It looks like it's actually:

1. Bios/UEFI
2. Grub/Systemd-boot
3. kernel
4. Initrd
5. Systemd (PID 1)

I thought initrd is only accessed at some point during PID 1, but it looks like it actually runs before PID1.

So far, I've learnt that initrd is distribution dependent and isn't directly part of our project.

**Comment by AdrianVovk** [comment]
I suggest reading [this man page](https://systemd.network/bootup.html). It will possibly help with your confusion about the boot process. But here's a quick summary of the initrd:

1. User presses power button on their device
2. BIOS/UEFI starts up (aka "firmware"; not under our control)
3. Firmware starts a bootloader (examples: GRUB, systemd-boot)
4. Bootloader finds and loads the kernel & initrd into memory
5. Bootloader hands control into the kernel, and tells it where the initrd is loaded
6. The kernel "mounts" the initrd as the root filesystem (it actually copies the files into a tmpfs, but its easier to think of it as mounting)
7. Once the kernel is ready to execute programs, it runs `/init`
8. `/init` does whatever it needs to find the real root filesystem and mount it somewhere (usually `/sysroot`)
9. `/init` tells the kernel to "pivot root" into /sysroot. This replaces / with everything that was mounted to /sysroot. In other words, from this point forwards the initrd's files are gone and now / is the _real_ root filesystem. `/init` keeps running from memory
10. `/init` finds and executes systemd in the real system. On UNIX `exec` stops executing one program and turns it into another. Effectively, at this moment `/init` stops exsiting and it is completely replaced by systemd from the real system

Many distros choose to use systemd inside of the initrd as well as outside of it. This is the case that you should care about. This is what is meant by "in the initrd" in the context of systemd. Here's what that looks like

1. User presses power button on their device
2. BIOS/UEFI starts up
3. Firmware starts a bootloader
4. Bootloader finds and loads the kernel & initrd into memory
5. Bootloader hands control into the kernel, and tells it where the initrd is loaded
6. The kernel "mounts" the initrd as the root filesystem
7. Once the kernel is ready to execute programs, it runs `/init`, which is actually systemd (I'll call this initrd-systemd)
8. initrd-systemd detects that it is running in the initrd, so it goes into initrd mode instead of a normal bootup
9. initrd-systemd executes units that find the real root filesystem and mount it to `/sysroot`
    - HERE is where you can include other services that do other things, like your battery check
    - Some of the units initrd-systemd executes load drivers that the kernel needs
    - One of the units initrd-systemd executes will by plymouth, which plays a boot animation and/or prompts for password
10. initrd-systemd cleans up after itself: most of the units that were running are stopped, but some can ask to stick around if important
11. Once /sysroot is ready, initrd-systemd then tells the kernel to pivot root. Again the initrd's files are now gone and / is replaced with whatever was mounted to /sysroot
12. initrd-systemd then searches the new root filesystem to find the real systemd (I'll call that real-systemd)
13. initrd-systemd replaces itself with real-systemd
14. real-systemd detects that it is _not_ running in an initrd, so it continues in normal bootup

**Comment by 1awesomeJ** [comment]
Wow, just like that, I mean you took out your precious time to explain this in such great details?

How awesome, how exciting.

Thank you so so so much @AdrianVovk.

I'm so so glad to be part of this community.

Thank you so so so much once again sir.

I'm super Thankful.

**Comment by 1awesomeJ** [comment]
Thank you so much sir.
It's really awesome to get your guidance.

**Comment by bluca** [comment]
> @bluca @poettering, Good evening sirs. For the task "detect battery level in initrd and if low refuse continuing to boot, print message and shut down"
> 
> I'm confused about the "in initrd" part.

There are various units that run only in the initrd, they generally are named "initrd-something", check the units/ directory. There's already a helper function to check whether battery is critically low, `battery_is_discharging_and_low()` in `src/shared/sleep-config.c`. A simple way to implement this task is to create a new simple tool that just calls that, modeled after `src/ac-power/ac-power.c` but that it exits with an error (and logs an emergency alert) if battery is critical and success otherwise, and add an initrd unit that calls it, which also sets FailureAction=poweroff-force - see: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#FailureAction=

**Comment by 1awesomeJ** [comment]
> > @bluca @poettering, Good evening sirs. For the task "detect battery level in initrd and if low refuse continuing to boot, print message and shut down"
> > I'm confused about the "in initrd" part.
> 
> There are various units that run only in the initrd, they generally are named "initrd-something", check the units/ directory. There's already a helper function to check whether battery is critically low, `battery_is_discharging_and_low()` in `src/shared/sleep-config.c`. A simple way to implement this task is to create a new simple tool that just calls that, modeled after `src/ac-power/ac-power.c` but that it exits with an error (and logs an emergency alert) if battery is critical and success otherwise, and add an initrd unit that calls it, which also sets FailureAction=poweroff-force - see: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#FailureAction=

Yay!!!! Nice to hear from you sir.

**Comment by 1awesomeJ** [comment]
> > @bluca @poettering, Good evening sirs. For the task "detect battery level in initrd and if low refuse continuing to boot, print message and shut down"
> > I'm confused about the "in initrd" part.
> 
> There are various units that run only in the initrd, they generally are named "initrd-something", check the units/ directory. There's already a helper function to check whether battery is critically low, `battery_is_discharging_and_low()` in `src/shared/sleep-config.c`. A simple way to implement this task is to create a new simple tool that just calls that, modeled after `src/ac-power/ac-power.c` but that it exits with an error (and logs an emergency alert) if battery is critical and success otherwise, and add an initrd unit that calls it, which also sets FailureAction=poweroff-force - see: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#FailureAction=

Awesome sir. Thank you so much, I'll get to work right away. Thank you sir.

**Comment by 1awesomeJ** [comment]
How has your day been sir? I look forward to a video meet with you soon. I hope my accent won't be terrible!
We were told to introduce ourselves to our community, Is there a place I can do that? I'm glad to have already interacted with @poettering , @keszybz , @medhefgo @mrc0mmand @yuwata @DaanDeMeyer @YHNdnzj. It's been a privilege.

**Comment by 1awesomeJ** [comment]
> > 
> > 
> > There are various units that run only in the initrd, they generally are named "initrd-something", check the units/ directory. There's already a helper function to check whether battery is critically low, `battery_is_discharging_and_low()` in `src/shared/sleep-config.c`. A simple way to implement this task is to create a new simple tool that just calls that, modeled after `src/ac-power/ac-power.c` but that it exits with an error (and logs an emergency alert) if battery is critical and success otherwise, and add an initrd unit that calls it, which also sets FailureAction=poweroff-force - see: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#FailureAction=


I want to create a new tool "battery-check.c" should I place it in "src/battery-check/battery-check.c" or could it just stay in "src/ac-power/battery-check.c"?

Also is the name choice good enough? or should I use something like "early-boot-battery-check.c"


I appreciate the time you commit to guiding me very very much, knowing how busy your schedules can be. It's a super awesome blessing to have you. Thank you sir.

Thank you everyone.

**Comment by 1awesomeJ** [comment]
Also, I'm not seeing any other use cases for this tool that may warrant adding the "--help", "--version" and "--verbose" flags as with the ac-power tool.

should I leave those out sir?

**Comment by 1awesomeJ** [comment]
Here's what I've done sir:
https://github.com/systemd/systemd/pull/27830

**Comment by 1awesomeJ** [comment]
@bluca 
Good evening sir,
I got an email from Outreachy that says I need to answer the question:
 "Do mentor(s) meet with the intern over phone or video chat?"

Should I answer no?

Or as I would like, would you permit to schedule a video call sir?

Thank you.

**Comment by bluca** [comment]
> @bluca Good evening sir, I got an email from Outreachy that says I need to answer the question: "Do mentor(s) meet with the intern over phone or video chat?"
> 
> Should I answer no?
> 
> Or as I would like, would you permit to schedule a video call sir?
> 
> Thank you.

We can continue using github PRs/issues for now, and re-evaluate if needed

**Comment by 1awesomeJ** [comment]
> > @bluca Good evening sir, I got an email from Outreachy that says I need to answer the question: "Do mentor(s) meet with the intern over phone or video chat?"
> > Should I answer no?
> > Or as I would like, would you permit to schedule a video call sir?
> > Thank you.
> 
> We can continue using github PRs/issues for now, and re-evaluate if needed

Okay sir.

While we wait for @poettering 's reviews on the open PR.

Could you give a guidance on the next feature implementation sir, just as you helped get me started on the current one.
I think these two should be next?
"audit error level logs that are relevant for early boot failures, and bump log level up"
"for each such log messages, have an entry in the message catalog, and that the entry has an URL"

**Comment by 1awesomeJ** [comment]
I'm currently studying journald for the purpose of our next features, I would appreciate any pointers on how to navigate.

I see the need to understand how these work:
1. Message Ids
2. Message logging and log levels.
3. Service activation by journald
4. QR encoding in C
5.socket activation in journald.

Thank you.

**Comment by poettering** [comment]
so i think we should have a small tool that uses the sd-journal APIs and initially looks for the most recent message in the journal that:

1. originates from the current boot
2. has a LOG_EMERG log level
3. originates from a system user

i.e. more or less what `journalctl -b -p emerg -o verbose -n1` shows you (if you invoke this on your system this hopefully returns an empty result of course), but then skipping over anythign where the _UID field doesn't pass uid_is_system(). if there is such a message it should open /dev/console, and take it over somehow (maybe just tell PID 1 to turn off status output via 
SIGRTMIN+21, so that it doesn't output anything onto it), empty the screeen probably via `\033[r`, see how vt_disallocate() does it. And then bring the contents of the message to screen in a very visible way (i.e. maybe even steal the "blue screen of death" feeling that mswindows has there, see https://en.wikipedia.org/wiki/Blue_screen_of_death#/media/File:Bsodwindows10.png – though we of course cannot fit that much into the terminal usually, because we have so few chars).

Next step would be encode the message in a QR code. QR codes have little space, so it the goal would be to put as much in as we can and the screen allows. Note that we have QR code support already in place in tools such as homectl, see src/shared/qrcode-util.h and its users. We shouldn#t try too hard with the QR code stuff though, it's a gimmick mostly I'd say. For example, we should probably not bother if the screen is just too small to carry it (i.e. traditional terminals have room for 80x25 chars, and that means could only fit a 80x50 qrcode on it at most, which is mighty little). it would be grand if we would use the TTY ioctls to determine terminal dimensions and then use a bigger QR code if the dimensions are larger (and thus stuff more info in it), and a smaller QR code if the dimensions are smaller (with less info then). It might even make sense to switch between regular human readable text display and the QR code every 5s or so, forth and back.

next question would be how to activate this service automatically when LOG_EMERG is seen, but let's talk about that later (need to feed some kids now), and I think this can be a separate step.

**Comment by 1awesomeJ** [comment]
> so i think we should have a small tool that uses the sd-journal APIs and initially looks for the most recent message in the journal that:
> 
> 1. originates from the current boot
> 2. has a LOG_EMERG log level
> 3. originates from a system user
> 
> i.e. more or less what `journalctl -b -p emerg -o verbose -n1` shows you (if you invoke this on your system this hopefully returns an empty result of course), but then skipping over anythign where the _UID field doesn't pass uid_is_system(). if there is such a message it should open /dev/console, and take it over somehow (maybe just tell PID 1 to turn off status output via SIGRTMIN+21, so that it doesn't output anything onto it), empty the screeen probably via `\033[r`, see how vt_disallocate() does it. And then bring the contents of the message to screen in a very visible way (i.e. maybe even steal the "blue screen of death" feeling that mswindows has there, see https://en.wikipedia.org/wiki/Blue_screen_of_death#/media/File:Bsodwindows10.png – though we of course cannot fit that much into the terminal usually, because we have so few chars).
> 
> Next step would be encode the message in a QR code. QR codes have little space, so it the goal would be to put as much in as we can and the screen allows. Note that we have QR code support already in place in tools such as homectl, see src/shared/qrcode-util.h and its users. We shouldn#t try too hard with the QR code stuff though, it's a gimmick mostly I'd say. For example, we should probably not bother if the screen is just too small to carry it (i.e. traditional terminals have room for 80x25 chars, and that means could only fit a 80x50 qrcode on it at most, which is mighty little). it would be grand if we would use the TTY ioctls to determine terminal dimensions and then use a bigger QR code if the dimensions are larger (and thus stuff more info in it), and a smaller QR code if the dimensions are smaller (with less info then). It might even make sense to switch between regular human readable text display and the QR code every 5s or so, forth and back.
> 
> next question would be how to activate this service automatically when LOG_EMERG is seen, but let's talk about that later (need to feed some kids now), and I think this can be a separate step.

Wow.
Thank you for the detailed breakdown sir.
It's awesome

**Comment by 1awesomeJ** [comment]
The PR for the first task still awaits your review though.

Looks like everybody else has their requests resolved sir.

Thank you.

**Comment by 1awesomeJ** [comment]
Although I'm lost on like 60% of the text, I'll study to bring myself up quickly.

It's exciting to know we have a qr code already in homectl.

It's more exciting to hear about your family sir.
Regards to them. I hope they love C and Python as much as you do sir.

**Comment by 1awesomeJ** [comment]
@bluca,
Good morning sir. Have you again visited the Shore? at Leith? You should host us there sometime so we can share the beauty.

Poettering has given guidance about us creating a tool that:
sd-journal APIs and initially looks for the most recent message in the journal that:

originates from the current boot
has a LOG_EMERG log level
originates from a system user


I have a few questions:

1. This tool will be returning just one message(the most recent based on the stated criteria) not a list of messages that match those criteria
2. If we then pass this message to our QR encoding tool, it should mean that when a user scans with their phone, It;s only one message they see (the most recent from he journal based on our filter criteria), not all the messages that could have caused a failure.
3. What if boot fails for more than one reason?

4. The tool should be in its own PR by itself?
5. I haven't quite found the sd_journal APIs or more accurately, I don't fully know what they are yet. But the what I picture is that this tool would use the following functions: ```sd_journal_open()``` to initialize the journal; three instances of ```sd_journal_add_match()``` to add each of our search criteria; ```sd_journal_seek_tail()``` to move to the most recent message.

Kindly guide on this sir.

6. The tool completely takes care of this requiremment: audit error level logs that are relevant for early boot failures, and bump log level up?

Thank you sir.

**Comment by bluca** [comment]
Yes, we can worry about that later, we'll give you some guidance. Work on the tool should be independent from this.

**Comment by 1awesomeJ** [comment]
> so i think we should have a small tool that uses the sd-journal APIs and initially looks for the most recent message in the journal that:
> 
> 1. originates from the current boot
> 2. has a LOG_EMERG log level
> 3. originates from a system user
> 
> i.e. more or less what `journalctl -b -p emerg -o verbose -n1` shows you (if you invoke this on your system this hopefully returns an empty result of course), but then skipping over anythign where the _UID field doesn't pass uid_is_system(). if there is such a message it should open /dev/console, and take it over somehow (maybe just tell PID 1 to turn off status output via SIGRTMIN+21, so that it doesn't output anything onto it), empty the screeen probably via `\033[r`, see how vt_disallocate() does it. And then bring the contents of the message to screen in a very visible way (i.e. maybe even steal the "blue screen of death" feeling that mswindows has there, see https://en.wikipedia.org/wiki/Blue_screen_of_death#/media/File:Bsodwindows10.png – though we of course cannot fit that much into the terminal usually, because we have so few chars).
> 


I love this blue screen of death idea. It's super cool. It'll be a big one for me to be part of the team that implements it.
I have interacted with journalctl, to know what each flag does, now i'm want to study the API functions more, I hope to have the PR open today, so that when you get back from your morning wlaks tomorrow, and have some minutes to spare, it'll be available for your initial reviews.

Thank you for choosing such as awesome project as this for my Internship. Thank you sirs.

**Comment by 1awesomeJ** [comment]
@bluca,
Good evening sir.
I'm considering the name: systemd-journal-latest-boot-message for our new tool.
I'm considering src/journal/journal-latest-boot-message.c for its location.

Kindly guide on these sir.
Thank you.

**Comment by 1awesomeJ** [comment]
Also, should it return a string or a journal object?

**Comment by poettering** [comment]
maybe call it systemd-bsod.service, as reference to this: https://en.wikipedia.org/wiki/Blue_screen_of_death

I think that would be funny but apt.

**Comment by poettering** [comment]
> 2\. If we then pass this message to our QR encoding tool, it should mean that when a user scans with their phone, It;s only one message they see (the most recent from he journal based on our filter criteria), not all the messages that could have caused a failure.

I guess. I don't think we have space in the QR code for much else. But of course, if it turns out we do, then we can add multiple

**Comment by 1awesomeJ** [comment]
> > 1. This tool will be returning just one message(the most recent based on the stated criteria) not a list of messages that match those criteria
> 
> i was thinking we should show the most recent one. @bluca suggested the oldest one of the boot. And he has some rason. Another option would be to simply iterate through it and show an updatign screen, i.e. refresh every 5s, cycling through all message sone by one and then starting from the top. But maybe don't bother with such things yet. for now, let's maybe stick to @bluca's suggestion and show the oldest one

😁😁 Alright sir.
But once we get that working, I can bother? With the iteration option?
I previously thought I would have to work on writing a qr function, but since we already have that, I should exert myself a little on the other tasks.

**Comment by 1awesomeJ** [comment]
> > 5. I haven't quite found the sd_journal APIs or more accurately, I don't fully know what they are yet. But the what I picture is that this tool would use the following functions: `sd_journal_open()` to initialize the journal; three instances of `sd_journal_add_match()` to add each of our search criteria; `sd_journal_seek_tail()` to move to the most recent message.
> 
> yeah, this sounds about right. (tail for most recent messages, head for oldest message)

Thank you sir.

**Comment by 1awesomeJ** [comment]
Good afternoon sirs.
While I continue working on testing the battery-check feature, I'd like to get started on the bsod.
The PR is here:
https://github.com/systemd/systemd/pull/28077

Kindly find time to review sirs.
Thank you.

**Comment by 1awesomeJ** [comment]
@poettering @bluca 
This should still be open?

Not that I want to push an opinion on it.

I just want to be sure I still get the privilege of getting my code reviewed by you both for the next 8.4 weeks!!!🙏🙏🙏

**Comment by bluca** [comment]
Yeah it closes automatically on merge but it's fine

**Comment by 1awesomeJ** [comment]
Ohh...
Thank you sir.

**Comment by 1awesomeJ** [comment]
@bluca,
Good evening sir.
For other boot failures, we're supposed to generate a qr code that users can scan with their phones to get help.

We made some progress on the journal tool today.
Currently we can fetch the first emergency message that originates from the current boot session for the current user, and then display it on a virtual terminal.

We now need to implement changing the tty colour to blue as well as include the qr code.

If it was the message that we needed to encode, it'll have meant we were close to finishing,
But the situation is that for any error message x, we need to encode a url that has a fix/useful info for error message x.

1. Are the URLs links to manpages, documentations, Stack Overflow, Google, GitHub issues or a combination?
2. Since we can't be sure what exactly error message x will be for all users, how do we make sure to generate the appropriate URL?

Is that where these come in
```
audit error level logs that are relevant for early boot failures, and bump log level up
for each such log messages, have an entry in the message catalog, and that the entry has an URL
```

Kindly guide on how I can approach this sir.

Perhaps, I'll be able to make some progress on it this weekend.

Thank you sir.

The journal tool to fetch the first emergency message is taking shape, although I have issues testing. I just want to get all the guidance I can from Poettering first, so that while he gets busy with other things, I can then work on testing locally and hopefully spotting areas of improvement. He did say the patch works on his own end though.

**Comment by poettering** [comment]
given the bsod tool is now quite advanced already, we need to think about how to trigger its activation.

So journald has a varlink IPC API already, see server_open_varlink() in src/journal/journald-server.c. I think best would be to build on that:

1. extend this varlink api with a new call io.systemd.Journal.Subscribe() or so, which should initially just take a single log level argument, and which will reply with a notificition reply every time a new message with the specified log level or higher is processed by journald. this notification would *not* contain the message payload though, clients are expected to check for that in the journal files directly.

2. the socket activation logic in systemd should be beefed up. Right now it (i.e. .socket units) knows ListenStream= only. We should add a new ConnectStream= setting. It should do the reverse of ListenStream=, i.e. not listen on some port and activate a service when a connection comes in, but instead connect to some socket, and activate a service once traffic is seen on the connection. The actual implementation of this should be relatively straight-forward I guess: if this mode is enabled just do socket()+connect() rather than socket()+bind()+listen(). 

3. The .socket units should at the same time gain one new setting WriteData= or so. It would take a base64 encoded blob of data, and all we do is write the data into the socket after binding it. 

4. Now we fit these parts together: we make systemd-bsod.service socket-activated via systemd-bsod.socket which contains ConnectStream=/run/systemd/journal/io.systemd.journal and WriteData= that contains the Varlink datagram in encoded form (i.e. just some brief json with a NUL byte, together encoded in base64). systemd-bsod will then be activated automatically whenever a LOG_ERMEGENCY message is processed by journald. sytemd-bsod will get the connected varlink socket passed, which it just reads the data off and throws it away.

**Comment by 1awesomeJ** [comment]
Wow, wow, wow...
Even if I only get your attention one day a week, it's always certainly worth more than I need.
I'm grateful sir.

Not everyone has it this good.

Thank you so so much.

**Comment by 1awesomeJ** [comment]
@bluca , a lot of what @poettering has graciously explained here still sound like modern latin if I'm being honest, I'll count on you to translate when needed sir. 😁😁.

You guys are the best.
Please consider streaming the All Systems Go event, some of us would like to watch you both.

Thank you.

**Comment by 1awesomeJ** [comment]
And certainly other members that may participate. We'll love to watch them too.

**Comment by 1awesomeJ** [comment]
@bluca,
Now, I need the first interpretation:
Extending ``` server_open_varlink()``` to  now include a new call ```io.systemd.Journal.Subscribe()```, adding a new ```ConnectStream= setting```, making the ```.socket units```  gain one new setting ```WriteData=```, and finally fitting these parts together to make systemd-bsod.service socket-activated via systemd-bsod.socket which contains ConnectStream=/run/systemd/journal/io.systemd.journal and WriteData= that contains the Varlink datagram in encoded form (i.e. just some brief json with a NUL byte, together encoded in base64)


Can these four be in different commits in thesame PR, or different PRs for each one, or could we group them into 2-3 PRs in all?

Thank you sir.

Sorry to bother on a weekend.

I intend getting my response whenever the time is good.
Thank you sir.

**Comment by 1awesomeJ** [comment]
It's exciting though how much i'll get familiar with the codebase, given all these deep concepts Poettering has mentioned. With the knowledge, I'll be able to fix at least one issue/TODO per month going forward.
I may also get to make meaningful comments on PRs that you geniuses open. It'll be awesome.

**Comment by 1awesomeJ** [comment]
I opened a PR to add a new call to ```server_open_varlink()``` here:
https://github.com/systemd/systemd/pull/28420

Not quite ready for your reviews, but i'd rather just get started and push as i gain more knowledge.

**Comment by 1awesomeJ** [comment]
> So journald has a varlink IPC API already, see server_open_varlink() in src/journal/journald-server.c. I think best would be to build on that:
> 
> 1. extend this varlink api with a new call io.systemd.Journal.Subscribe() or so, which should initially just take a single log level argument, and which will reply with a notificition reply every time a new message with the specified log level or higher is processed by journald. this notification would _not_ contain the message payload though, clients are expected to check for that in the journal files direc

@poettering, @bluca , Good morning sirs.

I've been trying to implement this, but I'm not quite clear on it.
I added the "Io.systemd.Journal.Subscribe" to the valink_server_bind_method_many() call, 
binding the call to a function I named vl_subscribe_var()

The implementation of this vl_subscribe_var() method is where I've gotten stuck.

I understand this method should take a single argument which is the log level a service wants to subscribe to: emerg, warning, debug etc.

But I'm not sure what it's prototype should be.

I have considered:
```static int vl_subscribe_var( Varlink *link, JsonVariant *parameters, VarlinkMethodFlags flags, void *userdata); ``` based on precedence in the codebase.
But: 

1. I'm not sure if all those parameters are needed

2. I haven't quite pictured how a service like systemd-bsod will call this API, what would it be calling and how, how does it pass the log level as parameters since most of these Varlink functions take many parameters that aren't even strings.
3. How does the vl_subscribe_var() fetch the log level and what does it do with it.


I would appreciate some sort of break down on this sirs.

**Comment by bluca** [comment]
Is the description above enough to get you started or do you need more details?

**Comment by 1awesomeJ** [comment]
😁😁😁
This is such an exciting question sir.
Thank you.

I haven't looked at the resource today, I'll probably do later in the night.

But I don't want to miss the chance of getting more guidance especially as we head into the weekend.

So yes sir I do need more details 😁😁

This is such an exciting question from you sir.
Thank you.

**Comment by 1awesomeJ** [comment]
@bluca 
I finally saw your face, and heard your voice here:
https://youtu.be/aygDMB23U2s

It's awesome 😁😁😁

**Comment by 1awesomeJ** [comment]
Two things sir:
Please call me by name - "Josh" a few times before we end this internship it'll be awesome.
2. About the ```--wait``` option, what I picture is if someone ran ```./systemd-bsod --wait```, then we'll invoke the ```sd_journal_wait()``` call.
3. But now that we're running it as with a unit, is there a way to pass that ```--wait``` option in the unit file? I only know of setting ```ExecStart=path```
But on taking another look now, I see we can actually do ```Execstart=systemd-bsod --wait```
4. So everytime journald processes a new log message, regardless of the log level, the call to ```sd_journal_wait()``` returns and the iteration continues, doesn't that impact our resources footprint, running that many iterations. I estimate a few thousands per boot session.
5. Am I over estimating?

**Comment by 1awesomeJ** [comment]
If I'm picturing this correctly, This approach does simplify things by a factor of 100!😂😂

I do hope though that @poettering still finds time to come guide on those complex approach. The knowledge from those four things he wants us to implement would be awesome to gain.

**Comment by 1awesomeJ** [comment]
@bluca,
Good evening sir.
I added the ```--wait``` option here:
https://github.com/systemd/systemd/pull/28697

kindly review at your convenience sir.

Thank you.

**Comment by 1awesomeJ** [comment]
@bluca,

Good morning sir.

Kindly provide guidance on these sir:

->audit error level logs that are relevant for early boot failures, and bump log level up
->for each such log messages, have an entry in the message catalog, and that the entry has an URL

1. They're outstanding tasks on my internship.
2. They address my concern of our QR code in systemd-bsod needing to contain more than just the string  already displayed. Having a url would be great.

I know your schedule may be very busy, But for as long as you can spare time to guide sir, I'll be available to refine these features(battery check and bsod) as much as possible.

Thank you sir.

**Comment by bluca** [comment]
> ->for each such log messages, have an entry in the message catalog, and that the entry has an URL

It's easier to start from this - you can look around for `log_emergency()` and `log_emergency_errno()` existing calls, and add a catalogue entry first. Then we can add a new page on docs/ so that it gets published on systemd.io and can list the same text as the catalogue entries

**Comment by 1awesomeJ** [comment]
@bluca, 
I've gone through "log.c" and "log.h".
I attempted to trace the variables passed to ```log_emergency()``` for instance to find where it gets used, so I could understand better how to add a catalogue entry, and how it'll be processed.

I traced the calls like so: ```log_emergency(r, "emergency : %m") ``` -> ```log_full(LOG_EMERG, r, "emergency: %m")``` (if PID =1),
then ```log_full_errno_zerook(LOG_EMERG, 0, r, "emergency: %m") ``` then log_internal()``` then ```log_internalv()``` then ```log_dispatch_internal()``` then ```write_to_journal()``` then ```log_do_header(), log_do_context(), and sendmsg()```


I didn't quite see any reference to the catalogue entry sir.

What I picture is that each message we process with ```log_emergency()``` and ```log_emergency_errno()``` should have some sort of message ID, which will be used to create an entry for them in the catalogue of all our journal messages, and so with this ID, we can fetch a particular message from the journal, and such a fetch would give us much details about the message including possible fixes to the problems they report.
This fetching from the catalogue could also somehow be done with an url, such that when that url is opened, a web page with full details on the emergency message is gotten.

But, first, I'm not even sure if what I picture is accurate as per your guidance sir. If it is, or close, how do I find the message catalogue, how do I amend existing ```log_emergency()``` calls to make sure those messages get pushed to the catalogue.

Sorry I'm verbose!

**Comment by 1awesomeJ** [comment]
Or could it be that you're not asking me to modify the existing calls to the logging functions, 
But rather just add the messages in those calls as entries in the message catalog.
I.e where there's a call like ```log_emergency(r, "Failed to locate shared memory") ```
What you're saying is that I should add the message "Failed to locate shared memory" as an entry in the journal catalog?
1. Is that right sir? If so:
2. How do I find the journal catalog to see it's current form?
3. How do I add new entries to it?

Thank you sir.

**Comment by 1awesomeJ** [comment]
I just noticed there are some functions in the sd-journal API for the message catalogue. I'll look them up.
Perhaps ```catalog_list_items()``` will be useful for fetching all existing entries?

**Comment by 1awesomeJ** [comment]
I see this entry:
-- f77379a8490b408bbe5f6940505a777b
Subject: The journal has been started
Defined-By: systemd
Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel

The system journal process has started up, opened the journal
files for writing and is now ready to process requests.

1. Does it mean every log message we generate should have the following fields: (i) SD_ID128 ID, (ii) Subject (iii) Defined-By (iv) support url (v) message description.

2. On the org.freedesktop webpage, the purpose of the catalog is stated as follows:
The message catalog has a number of purposes:

(a)Provide the administrator, user or developer with further information about the issue at hand, beyond the actual message text
(b) Provide links to further documentation on the topic of the specific message
(c) Provide native language explanations for English language system messages
(d) Provide links for support forums, hotlines or other contacts
If the fields listed in (1) above are all there are, then only (a) and (d) seem to be taken care of, (b) and (c) still need their fields added.

3. Does this mean I have to manually check all calls to ```log_emergency()``` and ```log_emergency_errno()``` to add entries for messages which currently do not have an entry in the catalog? How would I go about that?

4. Does this also mean that all those "A.C power messages"  from systemd-battery-check and log messages for Failed function calls that I've so far made, should also have their own entries in the journal ideally? Or would it be more expedient to just do for emergency level messages?

**Comment by 1awesomeJ** [comment]
This entry seems to be quite cool, it has everything but the native language explanations:
```
-- a596d6fe7bfa4994828e72309e95d61e
Subject: Messages from a service have been suppressed
Defined-By: systemd
Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
Documentation: man:journald.conf(5)

A service has logged too many messages within a time period. Messages
from the service have been dropped.

Note that only messages from the service in question have been
dropped, other services' messages are unaffected.

The limits controlling when messages are dropped may be configured
with RateLimitInterval= and RateLimitBurst= in
/etc/systemd/journald.conf. See journald.conf(5) for details.
```

**Comment by bluca** [comment]
Don't worry about translations, that is done separately. Catalog entries are added to catalog/systemd.catalog.in, you need to look at the emergency level messages, and for each one generate a new uuid with `systemd-id128 new` and add a catalog entry, and then change the logging call to include the id as a MESSAGE_ID so that they are tied together. This might require some modification to the log_emergency() and log_emergency_errno() functions

**Comment by 1awesomeJ** [comment]
> Don't worry about translations, that is done separately. Catalog entries are added to catalog/systemd.catalog.in, you need to look at the emergency level messages, and for each one generate a new uuid with `systemd-id128 new` and add a catalog entry, and then change the logging call to include the id as a MESSAGE_ID so that they are tied together. This might require some modification to the log_emergency() and log_emergency_errno() functions

Okay sir.
It should be fun.

I'll reach out whenever I'm stuck.

**Comment by 1awesomeJ** [comment]
@bluca,

Currently, I see 23 calls total to both ```log_emergency()``` and ```log_emergency_errno```

None of the messages are currently in "catalog/systemd.catalog.in"

This is the approach I'm considering:

step 1:
On my machine, I'll run ```systemd-id128 new``` I'll get an output like ```ba55201f2aaa47c7860460a14b13f457```

step 2:
I'll go to src/systemd/sd-messages.h, and add a new macro like so:
```#define SD_MESSAGE_FAILED_SELINUX_LOAD                    SD_ID128_MAKE(ba, 55, 20, 1f, 2a, aa, 47, c7, 86, 04, 60, a1, 4b, 13, f4, 57)```

step 3:
I'll go to catalog/system.catalogue.in and add an entry like so:
```
-- ba55201f2aaa47c7860460a14b13f457
Subject: Failed to load SELinux policy
Defined-By: systemd
Support: %SUPPORT_URL%

Systemd was unable to load the SELinux policy for a system that has "enforce" set to true,
You may try the following to fix:
1. Set SELinux status to "permissive";
2. Restore your file context;
3. Rebuild your SELinux Policy.
```
(The last part will take lots of inputs during the PR review).

Is that all I need do as far as adding catalog entries for the messages is concerned?

**Comment by bluca** [comment]
You also need to change log_emergency*() so that it uses a message-id when logging, and takes it as a parameter, so that each call can specify the ID you are adding
